### ISSUE TITLE
Integrating .NET Core Debugger in other extensions

### ISSUE THREAD

**Issue by lambdakris** [OmniSharp/omnisharp-vscode#1364]
Would it be possible for another extension to provide the ability to download and setup the .NET Core Debugger? For example, it would beneficial to users to be able to download and setup the .NET Core Debugger with the Ionide F# extension so that they would not need to install the C# extension just for the debugger. Would this be a scenario you would consider helping unblock?

**Comment by gregg-miskelly** [comment]
@caslan @rajkumar42 @weinand

### Technical Challenges

There are three technical challenges with VS Code as it exists today that make this scenario non-ideal:

1: Registration of the debug adapter in package.json - today the C# extension registered the 'coreclr' debug adapter. If another extension tried to also registered 'coreclr' things will not work very well. I don't know if vscode handles this at all, but it can't possibly work very well as there is actually a fair amount of json code for this as it declares the attach and launch schema, which is something that, so far at least, has changed relatively often. The alternative for F# to register a new debug adapter (ex: 'fsharp-coreclr'). This could potentially work if we make some tweaks to the debugger, but it means that F# needs to keep a copy of all the schema.
2: Versioning - every time the C# extension ships a new debugger, F# would need to do work to pick it up.
3: Duplication of platform detection and downloading code - this lousy but livable IMHO. The F# extension would just make a copy of the code from the C# extension. This code doesn't change all that often, so it probably will not be too hard for F# to stay up to date.

### Suggested solution

Se here is my suggestion - instead of trying to have the F# extension make a copy of the debugger, we try and make it so that F# users can easily know that the should install the C# extension. If there are already VS Code APIs for detecting and installing an extension, F# could use this with `vscode.window.showWarningMessage` to pop up a warning on activation if C# isn't installed. Another option would be for the VS Code debugger to have some basic knowledge of 'debug adapter name -> VS Code Extension' (like it does for 'programming language->VS Code Extension') and suggest users install C# if there is a request to debug and the 'coreclr' debug adapter is missing. A third option would be some sort of explicit dependency in package.json to tell VS Code to install the C# extension when installing F#.

There is a related problem of loading the F# EE when someone creates it. We can solve this by giving F# some way of telling the C# debugger where to load this EE from. It could be some sort of configuration file, or a launch.json attribute (example: `"debuggerExtensionDirectory":"~/.vscode/extensions/fsharp-*/ee"`).

**Comment by lambdakris** [comment]
Yeah, I think the technical challenges you point out are relevant. I too wondered about resolving whether another extension had already installed the debugger and it's also good to avoid duplicating effort. The idea of using extensionDependencies in the Extension Manifest file or something along those lines to bring in or prompt for the C# extension as needed is a good start. Is there any possibility of separating out the debugger into it's own package such that the .NET Core Debugger could be referenced as an extensionDependency by any .NET language without any one language taking on a dependency on the other?

**Comment by gregg-miskelly** [comment]
@lambdakris is there a reason you don't want to just depend on the C# extension? Granted that maybe it isn't quite the perfect solution, but the C# extension is pretty small, there is a fair amount of sharing in the core infrastructure of the extension, and it is certainly more covenant for my team to deal with only one extension / repo.

**Comment by lambdakris** [comment]
I can't think of a technical reason, and I along with others have installed the C# extension to enable debugging of .NET (including F#) code in VS Code. My suggestion is more towards a UX bend and a preference for a clear separation of concerns. Even if any objections about installing the C# extension to debug other .NET lang code can be rationalized away, I suspect there will be many "purists" (functional programming especially appeals to devs with a purist inclination) who would stumble on the idea of installing an extension for a language they are not looking to use in order to enable debugging for their intended language. I understand if at this early stage breaking out the debugger into it's own extension might be counter productive to the debugger effort as a whole and I certainly don't want that to happen, but if that is the case, perhaps the team might keep this idea in mind as things continue to develop?
